### PR TITLE
2-Git

### PR DESCRIPTION
**Pull request by dahachm**
**2-Git**

В папке **screenshots** - скриншоты шагов из домашней работы.

**README.md** - отчет о проделанной работе. 

**screenshots.tgz** - архив каталога screenshots.

@SagurovA93

### CODE REVIEW

File: 2-Git/README.md [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+
+   There is another way to replace HEAD to the first commit in the branch:
+
+   ```
+   $ git reset --hard HEAD~2
+   ```
+
+   ![Screenshot_17](screenshots/Screenshot_17.png)
+
+   The difference between using **checkout** and **reset** commands is that **checkout** replace HEAD pointer in *detached HEAD* state and **reset** update both HEAD and branch state.
```
Comment (SagurovA93) [inline 2-Git/README.md:127]: Great! That's exactly what I would ask you if there no comments :)

File: 2-Git/README.md [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+   ```
+   $ git remote add origin https://github.com/dahachm/lection_git_hw.git
+   ```  
+
+   ![Screenshot_18](screenshots/Screenshot_18.png)
+
+  ## 3. Push all branches to the remote repo
+    
+   ```
+   $ git push -u origin master
```
Comment (SagurovA93) [inline 2-Git/README.md:151]: How you can push all branches at once?

File: 2-Git/README.md [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    
+   ![Screenshot_57](screenshots/Screenshot_57.png)
+   ![Screenshot_58](screenshots/Screenshot_58.png)
+
+  ## 10. Protect develop branch to allow everyone to create Merge Request into it
+    
+   ![Screenshot_59](screenshots/Screenshot_59.png)
+   ![Screenshot_60](screenshots/Screenshot_60.png)
+
+  ## 11. Allow anyone do anything in branches like "feature-*"
```
Comment (SagurovA93) [inline 2-Git/README.md:251]: This is a minor, but `feature-*` is not the same as `feature*`. This is just to remind you to be very attentive.
No fixes required here.

File: 2-Git/README.md [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+   ![Screenshot_52](screenshots/Screenshot_52.png)
+
+  ## 8. Protect master branch to allow only maintainers to merge into it, and restrict all to push there
+    
+   ![Screenshot_53](screenshots/Screenshot_53.png)
+   ![Screenshot_54](screenshots/Screenshot_54.png)
+   ![Screenshot_55](screenshots/Screenshot_55.png)
+   ![Screenshot_56](screenshots/Screenshot_56.png)
+
+  ## 9. Protect release branches by wildcard (release-* for example) and allow only maintainers to merge into it
```
Comment (SagurovA93) [inline 2-Git/README.md:241]: The `release*` is not the same as `release-*`. 

No fixes required, but next time, please be more attentive.

**Review by SagurovA93 — APPROVED** [review]
That's very good work! Thank you for being diligent, keep going.

As additional questions I would like to ask:
1. What will happen when your local GitLab docker container will stop? Will it keep configurations and code? 
2. Can you `push -f` to the master branch as a repository maintainer? 
3. Why the Gitlab offers to visit the Merge Request with a such strange URI? 
image
How to change the port number to the actual one (9080) and the Gitlab hostname?

**Comment by dahachm** [comment]
@SagurovA93 , thank you a lot!


> 1. What will happen when your local GitLab docker container will stop? Will it keep configurations and code?

While the GitLab container is stopped, the local GitLab is not available until the container is started again. 
All the repos and configurations will remain the same and will be accessible again.

**Comment by dahachm** [comment]
> 2. Can you `push -f` to the master branch as a repository maintainer?

Tricky question :D

According to documentation, default branch is *Fully protected* (by default, this rule can be changed in Admin area, see the pecture below) and **no-one can force push** in it and I assumed my master branch as the default one.

Screenshot_2
    
But while testing it I found out that I'm **still able to force push in master branch as maintainer** (when it's not protected).

Same problem was discussed here: https://gitlab.com/gitlab-org/gitlab/-/issues/244151

So *if new repo was created empty, master\main branch won't be added to protected branch and default branch access rule won't apply*. To be honest, I'm still a bit confused about such behavior.

BUT I'm 100% sure, that if the branch added to **protected branches** manually **no-one can force push in it**.

In this task we already added master branch to protected, so my answer is: 
**No, I can't  `push -f` to the master branch as a repository maintainer , when it's protected.**

Screenshot_4

**Comment by dahachm** [comment]
> 3. Why the Gitlab offers to visit the Merge Request with a such strange URI? How to change the port number to the actual one (9080) and the Gitlab hostname?

Last time I started docker with GitLab I didn't specified container **hostname** and **port numbers** on host machine (so docker set it by itself).

To avoid this, I just had to run container with the following command:
    
    $ docker run -d --hostname my-gitlab -p 9080:80 -p 22 gitlab/gitlab-ce:latest
   
   
I also added "192.168.56.2   my-gitlab" (IP addr of my VM) to /etc/hosts file on my hots machine, so now I can access GitLab WebUI om "my-gitlab:9080" and to foolow the offered link to create new merge request I need to *add port 9080* ('cause my web browser is trying to connect to port 80 by default).

Screenshot_3

Screenshot_70

File: 2-Git/README.md (hunk 2 above)
Comment (dahachm) [reply 2-Git/README.md:151]: git push origin --all

File: 2-Git/README.md (hunk 4 above)
Comment (dahachm) [reply 2-Git/README.md:241]: Okay, I just thought that this is the place where we can choose wildcards by ourselves :)

File: 2-Git/README.md (hunk 2 above)
Comment (SagurovA93) [reply 2-Git/README.md:151]: yes

**Pull request merged**